// vite-react-ssg/dist/shared/vite-react-ssg.DTTf7HZy.mjs
import React from 'react';
import * as ReactDOM from 'react-dom';

const CopyReactDOM = {
  ...ReactDOM
};
const { version, render: reactRender, hydrate: reactHydrate } = CopyReactDOM;
const isReact18 = Number((version || "").split(".")[0]) > 17;
const isReact19 = Number((version || "").split(".")[0]) > 18;
function render(app, container, renderOptions = {}) {
  const { useLegacyRender } = renderOptions;
  if (useLegacyRender || !isReact18) {
    reactRender(app, container);
  } else if (isReact19) {
    import('react-dom/client').then(({ default: { createRoot } }) => {
      const root = createRoot(container);
      React.startTransition(() => {
        root.render(app);
      });
    });
  } else {
    CopyReactDOM.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.usingClientEntryPoint = true;
    const { createRoot } = CopyReactDOM;
    if (!createRoot) {
      throw new Error("createRoot not found");
    }
    const root = createRoot(container);
    CopyReactDOM.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.usingClientEntryPoint = false;
    React.startTransition(() => {
      root.render(app);
    });
  }
}
function hydrate(app, container, renderOptions = {}) {
  const { useLegacyRender } = renderOptions;
  if (useLegacyRender || !isReact18) {
    reactHydrate(app, container);
  } else if (isReact19) {
    import('react-dom/client').then(({ default: { hydrateRoot } }) => {
      React.startTransition(() => {
        hydrateRoot(container, app);
      });
    });
  } else {
    CopyReactDOM.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.usingClientEntryPoint = true;
    const { hydrateRoot } = CopyReactDOM;
    if (!hydrateRoot) {
      throw new Error("hydrateRoot not found");
    }
    React.startTransition(() => {
      hydrateRoot(container, app);
      CopyReactDOM.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.usingClientEntryPoint = false;
    });
  }
}

function documentReady(_passThrough) {
  if (document.readyState === "loading") {
    return new Promise((resolve) => {
      document.addEventListener("DOMContentLoaded", () => resolve(_passThrough));
    });
  }
  return Promise.resolve(_passThrough);
}

export { documentReady as d, hydrate as h, render as r };
