// vite-react-ssg/dist/shared/vite-react-ssg.BhGKpaqQ.mjs
function joinUrlSegments(a, b) {
  if (!a || !b)
    return a || b || "";
  if (a[a.length - 1] === "/")
    a = a.substring(0, a.length - 1);
  if (b[0] !== "/")
    b = `/${b}`;
  return a + b;
}
function removeLeadingSlash(str) {
  return str[0] === "/" ? str.slice(1) : str;
}
function stripBase(path, base) {
  if (path === base)
    return "/";
  const devBase = withTrailingSlash(base);
  return path.startsWith(devBase) ? path.slice(devBase.length - 1) : path;
}
function withTrailingSlash(path) {
  if (path[path.length - 1] !== "/")
    return `${path}/`;
  return path;
}
function withLeadingSlash(path) {
  if (path[0] !== "/")
    return `/${path}`;
  return path;
}

function convertRoutesToDataRoutes(routes, mapRouteProperties, parentPath = []) {
  return routes.map((route, index) => {
    const treePath = [...parentPath, String(index)];
    const id = typeof route.id === "string" ? route.id : treePath.join("-");
    route.id = id;
    if (isIndexRoute(route)) {
      const indexRoute = {
        ...route,
        ...mapRouteProperties(route),
        id
      };
      return indexRoute;
    } else {
      const pathOrLayoutRoute = {
        ...route,
        ...mapRouteProperties(route),
        id,
        children: void 0
      };
      if (route.children) {
        pathOrLayoutRoute.children = convertRoutesToDataRoutes(
          route.children,
          mapRouteProperties,
          treePath
          // manifest,
        );
      }
      return pathOrLayoutRoute;
    }
  });
}
function isIndexRoute(route) {
  return route.index === true;
}

export { withTrailingSlash as a, convertRoutesToDataRoutes as c, joinUrlSegments as j, removeLeadingSlash as r, stripBase as s, withLeadingSlash as w };
